src/concrete/ml/torch/lora.py

```diff
@@ -0,0 +1,326 @@
+        else:
+            # Weight is (out_features, in_features)
+            output = input_tensor @ self.weight.t()
+        if self.bias is not None:
+            output += self.bias
+        return output
+
+
+class BackwardModuleLinear(torch.nn.Module):
+    """Backward module for linear layers."""
+
+    def __init__(self, weight, weight_transposed=False):
+        super().__init__()
+        self.weight = weight
+        self.weight_transposed = weight_transposed
+
+    def forward(self, grad_output):
+        """Backward pass for linear layers.
+
+        Args:
+            grad_output: The gradient output tensor.
+
+        Returns:
+            The gradient input tensor after applying the backward pass.
+        """
+        if self.weight_transposed:
+            grad_input = grad_output @ self.weight.t()
+        else:
+            grad_input = grad_output @ self.weight
+        return grad_input
+
+
+class CustomLinear(torch.nn.Module):
+    """Custom linear module."""
+
+    def __init__(self, weight, bias=None, weight_transposed=False):
+        super().__init__()
+        self.forward_module = ForwardModuleLinear(weight, bias, weight_transposed)
+        self.backward_module = BackwardModuleLinear(weight, weight_transposed)
+
+    def forward(self, input_tensor):
+        """Forward pass of the custom linear module.
+
+        Args:
+            input_tensor: The input tensor.
+
+        Returns:
+            The output tensor after applying the custom linear module.
+        """
+        return ForwardBackwardModule.apply(input_tensor, self.forward_module, self.backward_module)
+
+
+class ForwardBackwardModule(torch.autograd.Function):
+    """Custom autograd function for forward and backward passes."""
+
+    @staticmethod
+    def forward(ctx, input_tensor, forward_module, backward_module):
+        """Forward pass of the custom autograd function.
+
+        Args:
+            ctx: The context object.
+            input_tensor: The input tensor.
+            forward_module: The forward module.
+            backward_module: The backward module.
+
+        Returns:
+            The output tensor after applying the forward pass.
+        """
+        ctx.backward_module = backward_module
+        output = forward_module.forward(input_tensor)
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        """Backward pass of the custom autograd function.
+
+        Args:
+            ctx: The context object.
+            grad_output: The gradient output tensor.
+
+        Returns:
+            The gradient input tensor after applying the backward pass.
+        """
+        backward_module = ctx.backward_module
+        grad_input = backward_module.forward(grad_output)
+
+        # grad_weight and grad_bias are not needed when computing the backward for LoRA
+        return grad_input, None, None
+
+
+def get_remote_names(model: torch.nn.Module, include_embedding_layers: bool = False) -> List[str]:
```

Comment:
this is too specific for a library function. e.g. for the `lm_head` moniker. 

I suggest you add two arguments `remote_layer_types` and `layer_reject_filter`. 

You can then call with `remote_layer_types = [nn.Linear, nn.Embedding, transformer.Conv1D]`. Thus CML does not need to know `TransformerConv1D`. and `layer_reject_filter = ['lm_head`]` 

You can thus remove the `include_embedding_layers` flag.

Reply:
but in the function we manually add `CustomLinear` if the user asks for `nn.Linear`.

Reply:
I am not sure here. This is very specific with the `replace_layers_with_custom`. I would not allow the user to select layers himself. For now we do linear layers in FHE and we have a workaround for lm_head and embedding as long as there are not fixed.